src/neural/cuda/layers.cc

```diff
@@ -1225,6 +1187,409 @@ ResidualBlock<DataType>::~ResidualBlock() {
+  wk_op_size_ = weights.ip3_pol_b.size();
+
+  encoder_heads_ = weights.pol_encoder_head_count;
+  policy_d_model_ = wq_op_size_;
+
+  allocAndUpload<DataType>(&ip_pol_w_, weights.ip_pol_w, scratch);
+  allocAndUpload<DataType>(&ip_pol_b_, weights.ip_pol_b, scratch);
+
+  allocAndUpload<DataType>(&ip2_pol_w_, weights.ip2_pol_w, scratch);
+  allocAndUpload<DataType>(&ip2_pol_b_, weights.ip2_pol_b, scratch);
+
+  allocAndUpload<DataType>(&ip3_pol_w_, weights.ip3_pol_w, scratch);
+  allocAndUpload<DataType>(&ip3_pol_b_, weights.ip3_pol_b, scratch);
+
+  allocAndUpload<DataType>(&ip4_pol_w_, weights.ip4_pol_w, scratch);
+
+  for (const auto& enc : weights.pol_encoder) {
+    EncoderWeights* pW = new EncoderWeights(enc, scratch);
+    encoder_weights_.emplace_back(pW);
+  }
+}
+
+template <typename DataType>
+AttentionPolicyHead<DataType>::EncoderWeights::EncoderWeights(
+    const LegacyWeights::EncoderLayer& cpu_weights, void* scratch) {
+  mha_q_size_ = cpu_weights.mha.q_b.size();
+  mha_k_size_ = cpu_weights.mha.k_b.size();
+  mha_v_size_ = cpu_weights.mha.v_b.size();
+  mha_dense_size_ = cpu_weights.mha.dense_b.size();
+  ffn_dense1_size_ = cpu_weights.ffn.dense1_b.size();
+  ffn_dense2_size_ = cpu_weights.ffn.dense2_b.size();
+
+  allocAndUpload<DataType>(&mha_q_w, cpu_weights.mha.q_w, scratch);
+  allocAndUpload<DataType>(&mha_q_b, cpu_weights.mha.q_b, scratch);
+
+  allocAndUpload<DataType>(&mha_k_w, cpu_weights.mha.k_w, scratch);
+  allocAndUpload<DataType>(&mha_k_b, cpu_weights.mha.k_b, scratch);
+
+  allocAndUpload<DataType>(&mha_v_w, cpu_weights.mha.v_w, scratch);
+  allocAndUpload<DataType>(&mha_v_b, cpu_weights.mha.v_b, scratch);
+
+  allocAndUpload<DataType>(&mha_dense_w, cpu_weights.mha.dense_w, scratch);
+  allocAndUpload<DataType>(&mha_dense_b, cpu_weights.mha.dense_b, scratch);
+
+
+  allocAndUpload<DataType>(&ln1_gammas, cpu_weights.ln1_gammas, scratch);
+  allocAndUpload<DataType>(&ln1_betas, cpu_weights.ln1_betas, scratch);
+
+  allocAndUpload<DataType>(&ffn_dense1_w, cpu_weights.ffn.dense1_w, scratch);
+  allocAndUpload<DataType>(&ffn_dense1_b, cpu_weights.ffn.dense1_b, scratch);
+
+  allocAndUpload<DataType>(&ffn_dense2_w, cpu_weights.ffn.dense2_w, scratch);
+  allocAndUpload<DataType>(&ffn_dense2_b, cpu_weights.ffn.dense2_b, scratch);
+
+  allocAndUpload<DataType>(&ln2_gammas, cpu_weights.ln2_gammas, scratch);
+  allocAndUpload<DataType>(&ln2_betas, cpu_weights.ln2_betas, scratch);
+}
+
+// taken from https://stackoverflow.com/questions/1659440/32-bit-to-16-bit-floating-point-conversion
+unsigned short float_to_half(const float x) {
+  const unsigned int b = (*(unsigned int*)&x) + 0x00001000;
+  const unsigned int e = (b & 0x7F800000) >> 23;
+  const unsigned int m = b & 0x007FFFFF;
+  return (b & 0x80000000) >> 16 |
+         (e > 112) * ((((e - 112) << 10) & 0x7C00) | m >> 13) |
+         ((e < 113) & (e > 101)) *
+             ((((0x007FF000 + m) >> (125 - e)) + 1) >> 1) |
+         (e > 143) * 0x7FFF;
+}
```

Comment:
Probably not important, but this gets round-to-even and infinity wrong. In dx12 have x86 code with a bit accurate fallback that can be moved to utils.